support named composite types

pH14 opened the pull request:
<!--

Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.

-->

As a building block for JSON sources, and to further close the gap with postgres compatibility, this PR adds in support for [named composite types](https://www.postgresql.org/docs/current/rowtypes.html).

These changes let us do some fun stuff like:

```
# create a composite type
materialize=> CREATE TYPE inventory_item AS (
    name            text,
    supplier_id     integer,
    price           numeric
);
CREATE TYPE

# use it in a table!
materialize=> CREATE TABLE on_hand (
    item      inventory_item,
    count     integer
);
CREATE TABLE

# insert rows into said table, coercing a ROW into our named type
materialize=> INSERT INTO on_hand VALUES (ROW('fuzzy dice', 42, 1.99), 1000);
INSERT 0 1

# select our type, among other columns
materialize=> SELECT * FROM on_hand;
          item          | count
------------------------+-------
 ("fuzzy dice",42,1.99) |  1000
(1 row)

# select individual columns from the type
materialize=> SELECT (item).supplier_id FROM on_hand;
 supplier_id
-------------
          42
(1 row)

# cast an unnamed Row into our type
materialize=> SELECT ROW('cookies', 100, 1.29)::inventory_item;
       record
--------------------
 (cookies,100,1.29)

# inspect the type
materialize=> SELECT pg_typeof(item) FROM on_hand;
   pg_typeof
----------------
 inventory_item
(1 row)
```

## Omissions

There are a few Postgres features related to named composite types that this PR omits. They might be worth adding in over time, but IMO they can be reasonably be done as separate PRs when needed. The features are:

### Omitted: inserts/updates to named fields

Postgres lets you individually set fields of composite types in insert/update statements:

```sql
create type complex as (r float8, i float8);
create type quad as (c1 complex, c2 complex);
create table quadtable(f1 int, q quad);

# not supported
insert into quadtable (f1, q.c1.r, q.c2.i) values(44,55,66);
# not supported
update quadtable set q.c1.r = 12 where f1 = 2;
```

Since most data will be coming in from sources, this one feels less important to solve right away.

### Omitted: functional notation

Postgres creates function names for each field of a composite type, so the following are equivalent:

```sql
SELECT (item).supplier_id FROM on_hand;
SELECT supplier_id(item) FROM on_hand;
```

This feature is also omitted from the PR, and only the more common access like `(item).supplier_id` is supported.

### Omitted: aliasing into anonymous composite type

Not sure if aliasing is the right word, but Postgres lets you read results into an anonymous composite type. Here's an example that creates an anonymous type `q`:

```sql
select q.a, q.b = row(2), q.c = array[row(3)], q.d = row(row(4)) from
    unnest(array[row(1, row(2), array[row(3)], row(row(4))),
                 row(2, row(3), array[row(4)], row(row(5)))])
      as q(a int, b record, c record[], d record);
```

This one is pretty cool, but IMO doesn't feel like a pre-req before introducing composite types / unblocking JSON sources.

## Testing

This PR includes a few types of tests:
* A port of Postgres's `rowtypes` tests (specifically the ones related to named composite types)
* A few quick pgwire tests to ensure we can deserialize named types into Records
* Upgrade/catalog tests to ensure we can read/write named composite types between restarts

### Motivation

Fixes https://github.com/MaterializeInc/materialize/issues/3516

<!--

Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug: [Link to issue.]

  * This PR adds a known-desirable feature: [Link to issue.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]

-->

### Tips for reviewer

As always, all feedback is welcome, be it Rust-isms, Postgres-isms, Materialize-isms or other. It's all still new to me 😄 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add support for Postgres's [named composite types](https://www.postgresql.org/docs/14/rowtypes.html) via `CREATE TYPE <name> AS (columns)` {{% gh 3516 %}}

benesch commented on src/repr/src/scalar.rs:
```diff
@@ -1497,6 +1497,8 @@ impl<'a> ScalarType {
                 value_type: t,
                 custom_oid,
             } => custom_oid.is_some() || t.is_custom_type(),
+            // TODO(phemberger): also need to check every field within the record
```
Hm, this whole `is_custom_type` thing seems pretty sketchy. Just reading up on https://github.com/MaterializeInc/materialize/blob/6a29b8604b058a95cae4761db7bafe5d9c5d9483/src/sql/src/func.rs#L533-L624 and I'm not entirely sure I understand what's going on.

sploiselle replied:
The idea is that we don't treat two custom types as interoperable, so you can't e.g. concatenate two custom lists together because we cannot determine the appropriate return type. lmk if you'd like to go through it in more depth, but that is essentially the shape of the hoop we're jumping through here.

pH14 commented:
Will leave polishing up the PR for proper review until next week, but it should be close now! We're passing a good % of the postgres rowtype tests, and it feels like the remaining compatibility could be left up to future PRs / demand. I think we should have enough to unblock JSON formatted sources with this

benesch commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -780,28 +781,58 @@ impl<T: AstInfo> AstDisplay for CreateTypeStatement<T> {
+            CreateTypeAs::Record => {
+                f.write_str("( ");
+
+                // TODO(phemberger): hmm do we create a separate field to store ColumnDefs directly?
```
Yeah, 100% we do. Seems to me like `CreateTypeAs` wants to look something like this:

```rust
enum CreateTypeAs {
    List { with_options: Vec<SqlOption<T>> },
    Map { with_options: Vec<SqlOption<T>> },
    Record { column_defs: Vec<ColumnDef<T>> }
}
```

pH14 replied:
Nice, that's what I've been leaning towards too. I'm a little scared that it'll cause a wide impact (I think there are several spots where it's helpful the Record column defs are treated as SqlOptions), so fingers crossed!

pH14 replied:
phew, wasn't bad and I think this helped tidy up the overall flow of things

benesch reviewed (commented):
:O this pretty much LGTM. Lemme know when you've done the polishing you mentioned and I can get you a real review!

benesch commented:
Btw, a super cool test to add to this PR would be to use the rust-postgres library to deserialize named composite types into Rust structs automatically: https://docs.rs/postgres-types/latest/postgres_types/index.html#composites. src/materialized/tests/pgwire.rs is the right home for  atest like that.

pH14 commented on src/sql-parser/tests/testdata/create:
```diff
@@ -18,84 +18,105 @@ CREATE TYPE custom AS MAP (key_type=text, value_type=bool)
 CREATE TYPE custom AS MAP ( key_type = text, value_type = bool )
 =>
-CreateType(CreateTypeStatement { name: UnresolvedObjectName([Ident("custom")]), as_type: Map, with_options: [DataType { name: Ident("key_type"), data_type: Other { name: Name(UnresolvedObjectName([Ident("text")])), typ_mod: [] } }, DataType { name: Ident("value_type"), data_type: Other { name: Name(UnresolvedObjectName([Ident("bool")])), typ_mod: [] } }] })
+CreateType(CreateTypeStatement { name: UnresolvedObjectName([Ident("custom")]), as_type: Map { with_options: [DataType { name: Ident("key_type"), data_type: Other { name: Name(UnresolvedObjectName([Ident("text")])), typ_mod: [] } }, DataType { name: Ident("value_type"), data_type: Other { name: Name(UnresolvedObjectName([Ident("bool")])), typ_mod: [] } }] } })
```
the structure of `CreateTypeStatement` changed a bit, so nearly all these lines are updating to the new format. At the end we add in a few lines to test the parsing of `CREATE TYPE foo AS (...)` statements

pH14 commented:
There are a few omissions (detailed in the PR summary), but this now covers Postgres's core named composite type tests. This is ready for real review :]

cc @benesch

pH14 commented:
Oh will also need to update our docs on https://materialize.com/docs/sql/create-type/ , will look into that tomorrow

sploiselle commented on src/repr/src/scalar.rs:
```diff
@@ -1411,6 +1411,24 @@ impl<'a> ScalarType {
     }
 
     /// Returns number of layers (akin to array dimensions) on a
```
Looks like this line got frankenstein'ed onto this code.

pH14 replied:
whoops, merge conflict resolution. good catch 🕵️‍♂️

sploiselle commented on src/repr/src/scalar.rs:
```diff
@@ -1448,6 +1466,29 @@ impl<'a> ScalarType {
+                    .collect_vec();
+                Record {
+                    fields: default_embedded_field_values,
+                    custom_name: custom_name.clone(),
```
I will need to think about this, but there's a chance that this violates the spirit of this function, in that it propagates the name/oid of the type, rather than its..."desaturated" value, which this function's meant to return.

pH14 replied:
I'm almost wondering if we don't want a `Record` match at all, and fall through to the default `v.clone()`. That's what happens for existing List/Map types with custom oids already.

If we don't propagate the name/oid at all, then `pg_typeof` will return a generic Row rather than the custom implementation, which breaks from postgres behavior

sploiselle commented on src/sql-parser/src/parser.rs:
```diff
@@ -758,6 +758,37 @@ impl<'a> Parser<'a> {
+                return self.expected(self.peek_pos(), "field name and type", self.peek_token());
+            }
+            if self.consume_token(&Token::Comma) {
+                // Continue.
```
ime there is usually some way of avoiding empty blocks like this; just a style thing, so nbd

benesch replied:
`Parser::parse_comma_separated` is probably the thing to use here.

benesch replied:
```rust
    fn parse_composite_type_definition(&mut self) -> Result<Vec<ColumnDef<Raw>>, ParserError> {
        self.expect_token(&Token::LParen)?;
        let fields = self.parse_comma_separated(|parser| {
            Ok(ColumnDef {
                name: parser.parse_identifier()?,
                data_type: parser.parse_data_type()?,
                collation: None,
                options: vec![],
            })
        })?;
        self.expect_token(&Token::RParen)?;
        Ok(fields)
    }
```

pH14 replied:
I initially cribbed it from `parse_columns` , but that's definitely the way!

sploiselle commented on src/sql/src/func.rs:
```diff
@@ -744,6 +744,21 @@ impl ParamList {
                         return None;
                     }
                 }
+                // As of today, with no UDFs, the only operators that accept records/row types as
```
Ah, this makes me pretty uncomfortable, though I hear you that without more complex impls, not super worthwhile. However, it's relatively few LOC to determine base equality of record types, so let me see if I can type up the implementation.

sploiselle commented:
@pH14 Looking at (and thinking about) this; would you mind looking at how PG handles different composite types in homogenizing functions like least/greatest? Seems like maybe they want to have their "customness" discarded in these operations

pH14 commented:
> Looking at (and thinking about) this; would you mind looking at how PG handles different composite types in homogenizing functions like least/greatest? Seems like maybe they want to have their "customness" discarded in these operations

Just to check, postgres does handle row-types in `least` and `greatest`:

```
postgres=# SELECT LEAST(ROW(0, 5), ROW(1,2), ROW(3, 4));
(0,5)
```

And will _not_ compare different types:

```
postgres=# create type pair as (a int, b int);
postgres=# SELECT LEAST(ROW(0, 5)::pair, ROW(1,2), ROW(3, 4));
ERROR:  LEAST types pair and record cannot be matched
```

But will compare equal named types:

```
postgres=# SELECT LEAST(ROW(0, 5)::pair, ROW(1,2)::pair, ROW(3, 4)::pair);
(0,5)

postgres=# SELECT pg_typeof(LEAST(ROW(0, 5)::pair, ROW(1,2)::pair, ROW(3, 4)::pair));
pair
```

Making this work will likely be a different unit of work, because we currently can't compare built-in Rows, let alone custom row types yet:

```
materialize=> SELECT LEAST(ROW(0, 5), ROW(1,2), ROW(3, 4));
ERROR:  least cannot be cast to uniform type: record(f1: integer,f2: integer) vs text
```

I believe this is the same issue as https://github.com/MaterializeInc/materialize/issues/10422

sploiselle commented:
@pH14 re: https://github.com/MaterializeInc/materialize/pull/10734#issuecomment-1049190832 that is perfect and exactly what I wanted to know/expected. Will review with that all in mind.

sploiselle reviewed (commented):
Do you mind taking a look at [this](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan/transform_ast.rs#L530-L541), and adjusting it in light of this PR? I think it can just be removed/error messages adjusted. Though, what this really points to is that when resolving polymorphic records, there is a failure step that ought to include checking record type lengths. So maybe add a TODO there for me?

pH14 commented on src/coord/src/catalog.rs:
```diff
@@ -2605,6 +2605,10 @@ impl ExprHumanizer for ConnCatalog<'_> {
                 self.humanize_scalar_type(value_type)
             ),
+            Record {
+                custom_name: Some(name),
```
I updated the PR to only set custom_oid when reading a custom record type from the catalog. I'll add a TODO that we want `custom_oid` and `custom_name` to really be an enum, so we can't accidentally set both

pH14 commented:
> Though, what this really points to is that when resolving polymorphic records, there is a failure step that ought to include checking record type lengths. So maybe add a TODO there for me?

Hmm would you mind pointing me to where this ought to happen? Not sure I have the full context. Outside of this TODO, I think I've addressed all of your comments! Thank you for the detailed review 🙇

pH14 commented:
> Do you mind taking a look at [this](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan/transform_ast.rs#L530-L541), and adjusting it in light of this PR? I think it can just be removed/error messages adjusted.

What did you have in mind for the error messages? I'm poking around with removing it, but I'm worried it'll snowball into a bigger change than this already is. Wondering if it makes sense to leave it in as a (somewhat small) optimization for comparisons over the built-in row type

sploiselle commented:
@pH14 Ah when I said "error messages adjusted" I meant in the test cases that will start failing. This is now just a duplicative check of the failure case when determining the common polymorphic type for binary functions. I think improving the error message is a bigger change, which you're right to avoid, but simply removing that code should "just work." lmk if I'm still being unclear and I can throw a diff your way.

pH14 commented:
👍 makes sense. I'm seeing some super strange errors after removing the transform... when loading https://github.com/MaterializeInc/materialize/blob/256945a79c91e73233b94ccbf6558263dc272aba/src/coord/src/catalog/builtin.rs#L1940-L1952 we get an obtuse error `WHERE clause error: no overload for unknown = unknown: = does not support implicitly casting from record(f1: pg_catalog.text?) to pg_catalog.text`. 

I'll dig into it further, but if I can't find a path forward within some amount of time, do you have any objections to leaving the transform code in? I think the problem here might actually be a separate parsing issue

sploiselle commented:
@pH14 Oh yikes--yeah! If it's not trivial to remove, then glad to handle in a follow-on.

benesch commented:
That transform is there for a reason! I don't think that it should be removed in this PR. I suspect it might need to be rewritten entirely in a different spot, to fix #10868, but that definitely feels like an entirely separate PR, unless I am misunderstanding something.

sploiselle commented:
@benesch The bit I'm referring to was simply the `Record` length checking

benesch commented:
Ahh, ok, @pH14's commit removed the whole thing!

pH14 commented:
Ahh just the length check! I don't think we can remove that in isolation -- if we do, then the transform will continue to rewrite the row comparison as a field-by-field comparison (only without a bounds check!) and we won't get a casting failure

sploiselle reviewed (approved):
LGTM mod nit and a note that merging this blocks merging #10734, which you've duly noted in that issue

pH14 commented on src/sql-parser/src/parser.rs:
```diff
@@ -2356,21 +2370,30 @@ impl<'a> Parser<'a> {
-        self.expect_token(&Token::LParen)?;
-        let with_options = self.parse_comma_separated(Parser::parse_data_type_option)?;
-        self.expect_token(&Token::RParen)?;
+        match self.expect_one_of_keywords(&[LIST, MAP]) {
```
good catch!
